formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Decoder.kt

```diff
@@ -0,0 +1,636 @@
+            && deserializer.descriptor == ByteArraySerializer().descriptor
+        ) {
+            @Suppress("UNCHECKED_CAST")
+            decoder.nextByteString(tags) as T
+        } else {
+            decodeByteArrayAsByteString = decodeByteArrayAsByteString || deserializer.descriptor.isInlineByteString()
+            super.decodeSerializableValue(deserializer)
+        }
+    }
+
+    override fun decodeString() = decoder.nextString(tags)
+
+    override fun decodeNotNullMark(): Boolean = !decoder.isNull()
+
+    override fun decodeDouble() = decoder.nextDouble(tags)
+    override fun decodeFloat() = decoder.nextFloat(tags)
+
+    override fun decodeBoolean() = decoder.nextBoolean(tags)
+
+    override fun decodeByte() = decoder.nextNumber(tags).toByte()
+    override fun decodeShort() = decoder.nextNumber(tags).toShort()
+    override fun decodeChar() = decoder.nextNumber(tags).toInt().toChar()
+    override fun decodeInt() = decoder.nextNumber(tags).toInt()
+    override fun decodeLong() = decoder.nextNumber(tags)
+
+    override fun decodeNull() = decoder.nextNull(tags)
+
+    override fun decodeEnum(enumDescriptor: SerialDescriptor): Int =
+        enumDescriptor.getElementIndexOrThrow(decoder.nextString(tags))
+
+    private fun isDone(): Boolean = !finiteMode && decoder.isEnd() || (finiteMode && readProperties >= size)
+
+    private fun verifyKeyTags(descriptor: SerialDescriptor, index: Int, tags: ULongArray?) {
+        if (cbor.verifyKeyTags) {
+            descriptor.getKeyTags(index)?.let { keyTags ->
+                if (!(keyTags contentEquals tags)) throw CborDecodingException("CBOR tags $tags do not match declared tags $keyTags for $descriptor")
+            }
+        }
+    }
+}
+
+internal class CborParser(private val input: ByteArrayInput, private val verifyObjectTags: Boolean) {
+    private var curByte: Int = -1
+
+    init {
+        readByte()
+    }
+
+    private fun readByte(): Int {
+        curByte = input.read()
+        return curByte
+    }
+
+    fun isEof() = curByte == -1
+
+    private fun skipByte(expected: Int) {
+        if (curByte != expected) throw CborDecodingException("byte ${printByte(expected)}", curByte)
+        readByte()
+    }
+
+    fun isNull() = (curByte == NULL || curByte == EMPTY_MAP)
+
+    fun nextNull(tags: ULongArray? = null): Nothing? {
+        processTags(tags)
+        if (curByte == NULL) {
+            skipByte(NULL)
+        } else if (curByte == EMPTY_MAP) {
+            skipByte(EMPTY_MAP)
+        }
+        return null
+    }
+
+    fun nextBoolean(tags: ULongArray? = null): Boolean {
+        processTags(tags)
+        val ans = when (curByte) {
+            TRUE -> true
+            FALSE -> false
+            else -> throw CborDecodingException("boolean value", curByte)
+        }
+        readByte()
+        return ans
+    }
+
+    fun startArray(tag: ULong) = startArray(ulongArrayOf(tag))
```

Comment:
`fun startArray(tag: ULong)`, ` fun startMap(tag: ULong)`, and `fun nextString(tag: ULong)` are used only in tests. Should we move them there as extensions?

Reply:
Makes sense. fixed in 26405e262fa800e399b810b01491ea6151cf01b4